ThunkLibs/libOpenCL/libOpenCL_Common.h

```diff
@@ -0,0 +1,229 @@
+#pragma once
+
+#include <cstring>
+#include <memory>
+#define CL_VERSION_3_0
+#define CL_USE_DEPRECATED_OPENCL_1_0_APIS
+#define CL_USE_DEPRECATED_OPENCL_1_1_APIS
+#define CL_USE_DEPRECATED_OPENCL_1_2_APIS
+#define CL_USE_DEPRECATED_OPENCL_2_0_APIS
+#define CL_USE_DEPRECATED_OPENCL_2_1_APIS
+#define CL_USE_DEPRECATED_OPENCL_2_2_APIS
+#define CL_TARGET_OPENCL_VERSION 300
+#include <CL/cl.h>
+
+namespace FEX::CL {
+  enum class AsyncCallbackType {
+    // Tied to context
+    TYPE_CLCREATECONTEXT,
+    TYPE_CLCREATECONTEXTFROMTYPE,
+    TYPE_CLSETCONTEXTDESTRUCTORCALLBACK,
+    TYPE_CLSETMEMOBJECTDESTRUCTORCALLBACK,
+    TYPE_CLSETPROGRAMRELEASECALLBACK,
+    TYPE_CLSETEVENTCALLBACK,
+    // Tied to cl_progam
+    TYPE_CLBUILDPROGRAM,
+    TYPE_CLCOMPILEPROGRAM,
+    TYPE_CLLINKPROGRAM,
+    // Tied to command_queue
+    TYPE_CLENQUEUENATIVEKERNEL,
+    TYPE_CLENQUEUESVMFREE,
+  };
+
+  union CallbackData {
+    struct _Header {
+      AsyncCallbackType Type;
+#ifdef GUEST_THUNK_LIBRARY
+      uintptr_t cb;
+#else
+      fex_guest_function_ptr cb;
+#endif
+    };
+  };
+
+  struct Type_SingleUserData {
+    CallbackData::_Header Header;
+
+    // User provided data that we need to keep around
+    void *user_data;
+
+    template <AsyncCallbackType Type>
+    static Type_SingleUserData Create(decltype(CallbackData::_Header::cb) a_0, void * a_1) {
+      static_assert(Type == AsyncCallbackType::TYPE_CLCREATECONTEXT ||
+        Type == AsyncCallbackType::TYPE_CLCREATECONTEXTFROMTYPE ||
+        Type == AsyncCallbackType::TYPE_CLSETCONTEXTDESTRUCTORCALLBACK ||
+        Type == AsyncCallbackType::TYPE_CLSETMEMOBJECTDESTRUCTORCALLBACK ||
+        Type == AsyncCallbackType::TYPE_CLSETPROGRAMRELEASECALLBACK ||
+        Type == AsyncCallbackType::TYPE_CLSETEVENTCALLBACK ||
+        Type == AsyncCallbackType::TYPE_CLBUILDPROGRAM ||
+        Type == AsyncCallbackType::TYPE_CLCOMPILEPROGRAM ||
+        Type == AsyncCallbackType::TYPE_CLLINKPROGRAM ||
+        Type == AsyncCallbackType::TYPE_CLENQUEUESVMFREE
+        ,
+        "Mismatch type for handling clProgram");
+
+      FEX::CL::Type_SingleUserData Data {
+        .Header {
+          .Type = Type,
+          .cb = a_0,
+        },
+        .user_data = a_1,
+      };
+
+      return Data;
+    };
+  };
+
+  struct Type_SingleUserDataCopy {
+    CallbackData::_Header Header;
+
+    // User provided data that we need to keep around
+    void *user_data;
+    size_t size;
+    ~Type_SingleUserDataCopy() {
+      free(user_data);
```

Comment:
Any particular reason for using malloc/free over new/delete?

Reply:
This is untyped data, the OpenCL function `clEnqueueNativeKernel` only gives us a `void* args` and `size_t cb_args`. Which by spec the driver needs to make a copy of.
Since this is untyped, we could change this over to `operator new` and `operator delete`, which feels more disgusting than just malloc+free.

Reply:
Why is new/delete more disgusting than malloc/free? Using the latter feels rather foreign in a C++ codebase.

Reply:
I switched it over already, but operator new and operator delete for opaque pointer is...weird.

Reply:
Ah, my bad for not seeing you updated the code already.

Why is it using `operator new` rather than `new`? Is that related to our allocator tricks?

Should `~Type_SingleUserDataCopy` be using `delete[]` instead of `delete`? Not sure if `operator delete(user_data);` is equivalent to the former.

Reply:
```
void *Get(size_t size) {
    return new void[size];
}
```
```
<source>:4:16: error: allocation of incomplete type 'void'
    return new void[size];
```

As far as I'm aware you can't do opaque new and delete.

Reply:
Welp, TIL. I've always just done `new char[size]` in these cases, but your approach is actually more correct.

Does the OpenCL spec say anything about data alignment, or is it alright that we potentially downgrade alignment here?

Reply:
Nothing in spec talks about alignment of this data. The data gets shuttled off to the GPU/OpenCL Device so likely ends up in a UBO and then the driver needs to deal with whatever BS the user passes in.